Index: src/main/groovy/text/SimpleTemplateEngine.java
===================================================================
--- src/main/groovy/text/SimpleTemplateEngine.java	(revision 6886)
+++ src/main/groovy/text/SimpleTemplateEngine.java	(working copy)
@@ -53,8 +53,14 @@
     }
 
     public Template createTemplate(Reader reader) throws CompilationFailedException, IOException {
+        return createTemplate(GroovyShell.class.getClassLoader(), reader);
+    }
+    
+    public Template createTemplate(ClassLoader parentLoader, Reader reader) 
+        throws CompilationFailedException, IOException 
+    {
         SimpleTemplate template = new SimpleTemplate();
-        GroovyShell shell = new GroovyShell();
+        GroovyShell shell = new GroovyShell(parentLoader);
         String script = template.parse(reader);
         if (verbose) {
             System.out.println("\n-- script source --");
