From 14bfcc5803db3abd5570f66ceb8278cb311fdc68 Mon Sep 17 00:00:00 2001
From: Jennifer Hickey <jennifer.hickey@springsource.com>
Date: Mon, 13 Dec 2010 15:43:47 -0800
Subject: [PATCH 1/3] Added commit only if new tx (to support nested @Transactional)

---
 .../transaction/NaiveDoubleTransactionManager.java |   30 ++++++++++---------
 1 files changed, 16 insertions(+), 14 deletions(-)

diff --git a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
index 60bd3c2..f18d251 100644
--- a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
+++ b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
@@ -23,20 +23,22 @@ public class NaiveDoubleTransactionManager implements PlatformTransactionManager
 
 	public void commit(TransactionStatus ts) throws TransactionException {
 		try {
-		final TransactionStatus tsb = copyTransactionStatus(status.get(ts));
-		try {
-			a.commit(ts);
-		}
-		catch (Throwable t) {
-			System.err.println("Continuing to commit tx despite this:" + t);
-		}
-		try {
-			b.commit(tsb);
-		}
-		catch (Throwable t) {
-			System.err.println("Can't commit tx" + t);
-			throw new TransactionException(t.getMessage(), t) {}; 
-		}
+    		if(ts.isNewTransaction()) {
+    		    final TransactionStatus tsb = copyTransactionStatus(status.get(ts));
+        		try {
+        			a.commit(ts);
+        		}
+        		catch (Throwable t) {
+        			System.err.println("Continuing to commit tx despite this:" + t);
+        		}
+        		try {
+        			b.commit(tsb);
+        		}
+        		catch (Throwable t) {
+        			System.err.println("Can't commit tx" + t);
+        			throw new TransactionException(t.getMessage(), t) {}; 
+        		}
+    		}
 		} finally {
 			status.remove(ts);
 		}
-- 
1.6.4.4


From d2ce39e706167e9814646383fbcf734e5f7b7fbd Mon Sep 17 00:00:00 2001
From: Jennifer Hickey <jennifer.hickey@springsource.com>
Date: Thu, 6 Jan 2011 10:40:53 -0800
Subject: [PATCH 2/3] Activate transaction synchronizations only after the second tx is committed

---
 .../transaction/NaiveDoubleTransactionManager.java |   22 ++++++++++++++-----
 1 files changed, 16 insertions(+), 6 deletions(-)

diff --git a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
index f18d251..b3456b3 100644
--- a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
+++ b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
@@ -24,9 +24,11 @@ public class NaiveDoubleTransactionManager implements PlatformTransactionManager
 	public void commit(TransactionStatus ts) throws TransactionException {
 		try {
     		if(ts.isNewTransaction()) {
-    		    final TransactionStatus tsb = copyTransactionStatus(status.get(ts));
+    		    final TransactionStatus tsa = copyTransactionStatusForFirst(ts);
+    		    final TransactionStatus tsb = copyTransactionStatusForSecond(status.get(ts));
+    		  
         		try {
-        			a.commit(ts);
+        			a.commit(tsa);
         		}
         		catch (Throwable t) {
         			System.err.println("Continuing to commit tx despite this:" + t);
@@ -44,11 +46,18 @@ public class NaiveDoubleTransactionManager implements PlatformTransactionManager
 		}
 	}
 
-	private TransactionStatus copyTransactionStatus(TransactionStatus ts) {
+	private TransactionStatus copyTransactionStatusForSecond(TransactionStatus ts) {
 		Object t = (ts instanceof DefaultTransactionStatus) ? ((DefaultTransactionStatus) ts).getTransaction() : null;
-		return new DefaultTransactionStatus(t,ts.isNewTransaction(), false,  false, false, null);
+		return new DefaultTransactionStatus(t,ts.isNewTransaction(), ((DefaultTransactionStatus)ts).isNewSynchronization(),  false, false, null);
 	}
 	
+	private TransactionStatus copyTransactionStatusForFirst(TransactionStatus ts) {
+        Object t = (ts instanceof DefaultTransactionStatus) ? ((DefaultTransactionStatus) ts).getTransaction() : null;
+        //Set new synchronization to false so trans sync won't kick in until 2nd tx committed
+        return  new DefaultTransactionStatus(t,ts.isNewTransaction(), false,  ((DefaultTransactionStatus)ts).isReadOnly(),  
+            ((DefaultTransactionStatus)ts).isDebug(),  ((DefaultTransactionStatus)ts).getSuspendedResources());
+    }
+	
 	public TransactionStatus getTransaction(TransactionDefinition td)
 			throws TransactionException {
 		TransactionStatus atx = a.getTransaction(td);
@@ -58,9 +67,10 @@ public class NaiveDoubleTransactionManager implements PlatformTransactionManager
 	}
 
 	public void rollback(TransactionStatus ts) throws TransactionException {
-		final TransactionStatus tsb = copyTransactionStatus(status.remove(ts));
+	    final TransactionStatus tsa = copyTransactionStatusForFirst(ts);
+        final TransactionStatus tsb = copyTransactionStatusForSecond(status.remove(ts));
 		try {
-			a.rollback(ts);
+			a.rollback(tsa);
 		}
 		catch (Throwable t) {
 			System.err.println("Continuing to rollback tx despite this:" + t);
-- 
1.6.4.4


From 4245606bcaf7f19e77a125251bd9de2fb048b808 Mon Sep 17 00:00:00 2001
From: Jennifer Hickey <jennifer.hickey@springsource.com>
Date: Wed, 26 Jan 2011 11:38:42 -0800
Subject: [PATCH 3/3] Tx map made thread-safe. Causing occasional NPEs and non-release of DS connections in multi-threaded env.

---
 .../transaction/NaiveDoubleTransactionManager.java |    3 ++-
 1 files changed, 2 insertions(+), 1 deletions(-)

diff --git a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
index b3456b3..a6631e9 100644
--- a/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
+++ b/spring-data-commons-core/src/main/java/org/springframework/persistence/transaction/NaiveDoubleTransactionManager.java
@@ -1,5 +1,6 @@
 package org.springframework.persistence.transaction;
 
+import java.util.Collections;
 import java.util.IdentityHashMap;
 import java.util.Map;
 
@@ -10,7 +11,7 @@ import org.springframework.transaction.TransactionStatus;
 import org.springframework.transaction.support.DefaultTransactionStatus;
 
 public class NaiveDoubleTransactionManager implements PlatformTransactionManager {
-	Map<TransactionStatus,TransactionStatus> status=new IdentityHashMap<TransactionStatus, TransactionStatus>();
+	Map<TransactionStatus,TransactionStatus> status=Collections.synchronizedMap(new IdentityHashMap<TransactionStatus, TransactionStatus>());
 	private final PlatformTransactionManager a;
 	
 	private final PlatformTransactionManager b;
-- 
1.6.4.4

