From a2d2861896838f1ed7d0a2ea1bec9ae40045437b Mon Sep 17 00:00:00 2001
From: Jennifer Hickey <jennifer.hickey@springsource.com>
Date: Mon, 31 Jan 2011 16:11:54 -0800
Subject: [PATCH] Temp hack to remove getValue()'s call to get id if a tx is running (since we always initialize the id by calling getId() immediately after a new object is persisted). Profiling results showed the isAnnotationPresent reflection call was causing huge bottlenecks.

---
 .../DetachableEntityStateAccessors.java            |    2 +-
 .../NodeEntityStateAccessorsFactory.java           |    2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/DetachableEntityStateAccessors.java b/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/DetachableEntityStateAccessors.java
index 53faf90..0c3a512 100644
--- a/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/DetachableEntityStateAccessors.java
+++ b/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/DetachableEntityStateAccessors.java
@@ -83,7 +83,7 @@ public class DetachableEntityStateAccessors<ENTITY extends GraphBacked<STATE>, S
 
     @Override
     public Object setValue(final Field field, final Object newVal) {
-        if (!transactionIsRunning()) {
+        if (!transactionIsRunning()  || PartialNodeEntityStateAccessors.getId(getEntity(),getEntity().getClass()) == null) {
             final ENTITY entity = getEntity();
             if (!isDirty(field) && isWritable(field)) {
                 Object existingValue;
diff --git a/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/NodeEntityStateAccessorsFactory.java b/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/NodeEntityStateAccessorsFactory.java
index 3ce36ad..35eca5f 100644
--- a/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/NodeEntityStateAccessorsFactory.java
+++ b/spring-data-neo4j/src/main/java/org/springframework/data/graph/neo4j/fieldaccess/NodeEntityStateAccessorsFactory.java
@@ -40,7 +40,7 @@ public class NodeEntityStateAccessorsFactory {
                     new PartialNodeEntityStateAccessors<NodeBacked>(null, entity, entity.getClass(), graphDatabaseContext, finderFactory), graphDatabaseContext) {
                 @Override
                 protected boolean transactionIsRunning() {
-                    return super.transactionIsRunning() && getId(entity, entity.getClass()) != null;
+                    return super.transactionIsRunning();
                 }
             };
         } else {
-- 
1.6.4.4

