Index: src/org/hibernate/dialect/PostgreSQLDialect.java
===================================================================
--- src/org/hibernate/dialect/PostgreSQLDialect.java	(revision 14408)
+++ src/org/hibernate/dialect/PostgreSQLDialect.java	(working copy)
@@ -129,6 +129,10 @@
 		return "create sequence " + sequenceName; //starts with 1, implicitly
 	}
 
+    public boolean supportsDeferrableConstraints() {
+        return true;
+    }
+
 	public String getDropSequenceString(String sequenceName) {
 		return "drop sequence " + sequenceName;
 	}
